File: programmers/algorithm_data_structure/dynamic_programming/서울에서 경산까지.py
def my_solution(K, travel):
    '''    
    전부 도보로 이동한다고 기본값을 설정하고
    각 경로를 자전거 이동으로 바꾸는 개념으로 생각

    dp => {자전거 이동으로 바꾼 만큼 감소할 시간 총량: 왼쪽 값 기준에서 감소하는 총 모금액의 최솟값}
    w_t, w_m: time on foot
    b_t, b_m: time, money by bicycle

    dictionay 자료형을 이용해 indexing 시간을 최소화
    매번 K 조건 검사하고 dp에서 제외하는 것보다
    나중에 한 번에 빼는 게 더 속도가 빠르다.
    (다른 문제에서도 똑같이 적용되는지는 봐야함)
    '''
    dp = {0: 0}

    for w_t, w_m, b_t, b_m in travel:
        items = list(dp.items())  # 에러 방지를 위해 list 처리
        for key, value in items:
            new_key, new_value = key + (w_t-b_t), value + (w_m-b_m)

            if new_key not in dp:
                dp[new_key] = new_value
            else:
                dp[new_key] = min(dp[new_key], new_value)

    MAX_MONEY = sum([tr[1] for tr in travel])
    MAX_TIME = sum([tr[0] for tr in travel])

    valid_list = [v for i, v in dp.items() if MAX_TIME-i <= K]
    return MAX_MONEY - min(valid_list)


def best_solution(K, travel):
    '''
    개념은 똑같다. 각각 k 값에서 가지는 최대값을 활용했다.
    이전 단계 k값에서 가지는 최대값을 활용하는 점화식이다.

    다만 실행 시간은 내 풀이가 5배 정도 빠르다.
    '''
    n = len(travel)

    memo = [[0 for j in range(K+1)] for x in range(n+1)]

    for i in range(1, n+1):
        t_walk, v_walk, t_bike, v_bike = travel[i-1]

        for j in range(K+1):
            walk = memo[i-1][j-t_walk]+v_walk if j>=t_walk and memo[i-1][j-t_walk]!=-1 else -1
            bike = memo[i-1][j-t_bike]+v_bike if j>=t_bike and memo[i-1][j-t_bike]!=-1 else -1

            memo[i][j] = max(walk, bike)

    return  memo[n][K]

File: programmers/algorithm_data_structure/dynamic_programming/서울에서_경산까지.java
package algorithm_data_structure.dynamic_programming;

public class 서울에서_경산까지 {

    /*
    K : 걸리는 시간의 합의 최댓값 제한
    travel: 
    [도보 이동에 걸리는 시간, 도보 이동 시 얻을 모금액, 
    자전거 이동에 걸리는 시간, 자전거 이동 시 얻을 모금액]

    (생각의 흐름대로 정리)
    동적프로그래밍으로 최댓값을 보장하면서 나아가려면
    점화식이 필요하며, 각 단계마다 최댓값이라는 게 인정되고,
    이전 단계의 최댓값을 활용해 현재 단계의 선택지(도보 또는 자전거: 2개) 중
    1개를 선택해야 함

    각 지점을 순차적으로 방문하며 살펴보자
    우선 첫 지점에서는 K 값 제한이 가능한 선에서 도보, 자전거 순으로 선택하면 된다.
    이 개념을 확장하여

    n번째 지점에서는 도보, 자전거 선택이 가능하며, 도보를 선택했을 때 최댓값은
    n-1 번째 지점에서 K - (도보 시간) 제한일 때 최댓값에 n 번째 지점의 돈을 더한 값이다.
    (K를 꽉 채움)
    자전거도 마찬가지이며 두 선택 중 최댓값을 고르면 된다.

    K 숫자가 높아 dp 배열이 커보이지만 
    불필요한 부분을 제외하고 돌리면 그렇게 많지도 않다.
    */
    public int mySolution(int K, int[][] travel) {
        int answer = 0;
        int entryIdx = 0;
        int[][] dp = new int[travel.length+1][K+1];

        for (int i = 1; i < dp.length; i++) {
            int[] curr = travel[i-1];
            entryIdx += curr[2];

            for (int j = entryIdx; j <= K; j++) {
                dp[i][j] = Math.max(
                    (j >= curr[0]) ? (dp[i-1][j - curr[0]] + curr[1]) : 0,
                    dp[i-1][j - curr[2]] + curr[3]
                );
            }
        }

        answer = dp[dp.length - 1][K];
        return answer;
     }
}